File: app/src/main/java/droiddevelopers254/droidconke/HomeActivity.java
package droiddevelopers254.droidconke;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.bumptech.glide.request.RequestOptions;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.hdodenhof.circleimageview.CircleImageView;
import droiddevelopers254.droidconke.ui.BottomNavigationBehaviour;
import droiddevelopers254.droidconke.views.activities.AuthenticateUserActivity;
import droiddevelopers254.droidconke.views.activities.EventFeedback;
import droiddevelopers254.droidconke.views.fragments.InfoFragment;
import droiddevelopers254.droidconke.views.fragments.MapFragment;
import droiddevelopers254.droidconke.views.fragments.ScheduleFragment;

import static droiddevelopers254.droidconke.utils.SharedPref.PREF_NAME;

public class HomeActivity extends AppCompatActivity {

    @BindView(R.id.accountImg)
    CircleImageView accountImg;
    @BindView(R.id.toolbarTitleText)
    TextView toolbarTitleText;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.container)
    ConstraintLayout container;
    @BindView(R.id.content_home)
    FrameLayout contentHome;
    @BindView(R.id.navigation)
    BottomNavigationView navigation;
    SharedPreferences sharedPreferences;
    public static int navItemIndex = 1; //controls toolbar titles and icons
    Button signInBtn;
    FirebaseAuth auth;
    FirebaseRemoteConfig firebaseRemoteConfig;
    public static boolean fabVisible = true;
    AppBarLayout.LayoutParams params;

    private GoogleSignInClient mGoogleSignInClient;


    private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = item -> {
        switch (item.getItemId()) {
            case R.id.navigation_home:
                navItemIndex = 0;

                //hide views not required by these fragment
                accountImg.setVisibility(View.GONE);
                toolbarTitleText.setText(R.string.info_title);

                //activate the hide toolbar
                params = (AppBarLayout.LayoutParams) toolbar.getLayoutParams();
                params.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                        | AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS);

                InfoFragment infoFragment = new InfoFragment();
                getSupportFragmentManager().beginTransaction().replace(R.id.content_home, infoFragment)
                        .commit();
                return true;
            case R.id.navigation_schedule:

                navItemIndex = 1;

                //hide views not required by these fragment
                accountImg.setVisibility(View.VISIBLE);
                toolbarTitleText.setText(R.string.schedule_title);

                //activate the hide toolbar
                params = (AppBarLayout.LayoutParams) toolbar.getLayoutParams();
                params.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                        | AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS);

                ScheduleFragment scheduleFragment = new ScheduleFragment();
                getSupportFragmentManager().beginTransaction().replace(R.id.content_home, scheduleFragment)
                        .commit();
                return true;
            case R.id.navigation_map:

                navItemIndex = 2;

                //hide views not required by these fragment
                accountImg.setVisibility(View.GONE);
                toolbarTitleText.setText(R.string.map_title);

                //show toolbar if it was hidden
                params = (AppBarLayout.LayoutParams) toolbar.getLayoutParams();
                params.setScrollFlags(0);  // clear all scroll flags

                MapFragment mapFragment = new MapFragment();
                getSupportFragmentManager().beginTransaction().replace(R.id.content_home, mapFragment)
                        .commit();
                return true;
        }
        return false;
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ButterKnife.bind(this);

        sharedPreferences = getSharedPreferences(PREF_NAME, MODE_PRIVATE);
        auth = FirebaseAuth.getInstance();
        //setup defaults for remote config
        firebaseRemoteConfig = FirebaseRemoteConfig.getInstance();

        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();

        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) navigation.getLayoutParams();
        layoutParams.setBehavior(new BottomNavigationBehaviour());

        if (auth.getCurrentUser() != null) {
            //load user profile image
            Glide.with(getApplicationContext()).load(auth.getCurrentUser().getPhotoUrl())
                    .thumbnail(Glide.with(getApplicationContext()).load(auth.getCurrentUser().getPhotoUrl()))
                    .apply(new RequestOptions()
                    .diskCacheStrategy(DiskCacheStrategy.ALL))
                    .transition(new DrawableTransitionOptions()
                    .crossFade())
                    .into(accountImg);
        }

        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);
        navigation.setSelectedItemId(R.id.navigation_schedule);


    }
    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_sign_out, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {

            FirebaseAuth.getInstance().signOut();

            //google sign out
            mGoogleSignInClient.signOut().addOnCompleteListener(this,
                    task -> {
                    });
            startActivity(new Intent(this, AuthenticateUserActivity.class));
            finish();

            return true;
        }else if(id==R.id.action_feedback){

            startActivity(new Intent(this, EventFeedback.class));
        }

        return super.onOptionsItemSelected(item);

    }
}

File: app/src/main/java/droiddevelopers254/droidconke/HomeActivity.kt
package droiddevelopers254.droidconke

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation
import androidx.navigation.findNavController
import androidx.navigation.ui.NavigationUI
import androidx.navigation.ui.setupActionBarWithNavController
import com.google.android.material.appbar.AppBarLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import droiddevelopers254.droidconke.ui.authentication.AuthenticateUserActivity
import droiddevelopers254.droidconke.ui.feedback.EventFeedbackActivity
import droiddevelopers254.droidconke.utils.SharedPref.PREF_NAME
import droiddevelopers254.droidconke.utils.SharedPref.TOKEN_SENT
import droiddevelopers254.droidconke.viewmodels.HomeViewModel
import kotlinx.android.synthetic.main.activity_home.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel

class HomeActivity : AppCompatActivity() {
    private val sharedPreferences: SharedPreferences by lazy { getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE) }
    private val firebaseRemoteConfig: FirebaseRemoteConfig by inject()
    private val firebaseMessaging: FirebaseMessaging by inject()
    private val firebaseAuth: FirebaseAuth by inject()
    private lateinit var params: AppBarLayout.LayoutParams
    private val homeViewModel: HomeViewModel by viewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        setSupportActionBar(toolbar)

        setupNavigation()
        setupActionBarWithNavController(findNavController(R.id.navHostFragment), drawer_layout)

        setupNotifications()

        //observe live data emitted by view model
        observeLiveData()

    }

    private fun setupNotifications() {
        firebaseMessaging.subscribeToTopic("all")

        if (BuildConfig.DEBUG) {
            firebaseMessaging.subscribeToTopic("debug")
        }
    }

    private fun setupNavigation() {
        val navController = Navigation.findNavController(this, R.id.navHostFragment)
        // Update action bar to reflect navigation
        NavigationUI.setupActionBarWithNavController(this, navController, drawer_layout)
        NavigationUI.setupWithNavController(nav_view, navController)
        navController.addOnDestinationChangedListener { _, destination, _ ->
            when (destination.id) {
                R.id.sessionDetailsFragment -> {
                    toolbar.visibility = View.GONE
                }
                else -> toolbar.visibility = View.VISIBLE
            }
        }

    }

    private fun observeLiveData() {
        homeViewModel.getUpdateTokenResponse().observe(this, Observer {
            when {
                it -> {
                    sharedPreferences.edit().putInt(TOKEN_SENT, 1).apply()
                }
                else -> {
                    sharedPreferences.edit().putInt(TOKEN_SENT, 0).apply()
                }
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_sign_out, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> {
                unsubscribeNotifications()
                firebaseAuth.signOut()
                startActivity(Intent(this, AuthenticateUserActivity::class.java))
                finish()
                return true
            }
            R.id.action_feedback -> {
                startActivity(Intent(this, EventFeedbackActivity::class.java))
            }
            else -> return super.onOptionsItemSelected(item)
        }
        return super.onOptionsItemSelected(item)
    }

    private fun unsubscribeNotifications() = lifecycleScope.launch {
        firebaseMessaging.unsubscribeFromTopic("all").await()
        if (BuildConfig.DEBUG) {
            firebaseMessaging.unsubscribeFromTopic("debug").await()
        }

        // TODO Add unsubscription from favourites
    }

    override fun onBackPressed() {
        val drawerLayout: DrawerLayout = findViewById(R.id.drawer_layout)
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }


    override fun onSupportNavigateUp() =
            Navigation.findNavController(this, R.id.navHostFragment).navigateUp()

    companion object {
        var navItemIndex = 1 //controls toolbar titles and icons
        var fabVisible = true
    }
}
